src/App.js

```diff
@@ -29,37 +30,103 @@ const App = () => {
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
+  // const [winner, setWinner] = useState(null);
+  // const [gameEnd, setGameEnd] = useState(false);
 
-  // Wave 2
+  const squareClicked =(id)=>{
+    let didPlay = false 
+    if (winner !== null) {return}
+    const newState = squares.map(row =>row.map(pos =>{
+      if (pos.id !== id) {return pos;}
+      if (pos.value !== '') {return pos;} 
+  //we know the id matches
+      didPlay = true
+      const newPos = {...pos}
+      newPos.value = currentPlayer
+      return newPos 
+    }))
+    if (didPlay) {
```

Comment:
👍 
Only update the state if the player ended up making a valid move. We are able to declare `didPlay` in our click handler, but modify it within the anonymous functions passed to map, due to closure semantics. `didPlay` is defined in the execution context of the map functions, so they can see and modify it. This lets us detect whether a move occurred here.